rust/src/io/reader.rs

```diff
@@ -183,22 +187,52 @@ impl<'a> FileReader<'a> {
     /// Read a batch of data from the file.
     ///
     /// The schema of the returned [RecordBatch] is set by [`FileReader::schema()`].
-    pub async fn read_batch(&self, batch_id: i32) -> Result<RecordBatch> {
+    pub(crate) async fn read_batch(
+        &self,
+        batch_id: i32,
+        params: impl Into<ReadBatchParams>,
+    ) -> Result<RecordBatch> {
         let schema = self.projection.as_ref().unwrap();
-        let batch_offset = self
-            .metadata
-            .get_offset(batch_id)
-            .ok_or_else(|| Error::IO(format!("batch {} does not exist", batch_id)))?;
-        read_batch(
-            &self.object_reader,
-            schema,
-            batch_id,
-            &self.page_table,
-            self.fragment_id,
-            batch_offset,
-            self.with_row_id,
-        )
-        .await
+        read_batch(self, &params.into(), schema, batch_id, self.with_row_id).await
+    }
+
+    /// Take by records by indices within the file.
+    ///
+    ///
+    pub async fn take(&self, indices: &[u32]) -> Result<RecordBatch> {
+        let mut indices = Vec::from(indices);
+        indices.sort();
+        let mut indices_per_batch: BTreeMap<usize, Vec<u32>> = BTreeMap::new();
```

Comment:
might be worth creating like BatchOffsets and BatchOffsetMap structs?

So here you can do something like 
```
stream::iter(self.metadata.indices_to_batches(indices))
.then(|(batch)| async move { self.read_batch(batch.id as i32, batch.offsets().as_slice()).await })
```
?

Reply:
Sure will do.